### PR TITLE
Moved old examples and added LinearRegressionExample

### PR DESCRIPTION
**Pull request by zaleslaw**
Hi, @karllessard I've added a simple example with linear regression, I hope it could be useful for newbies with Java TF API because it's more shorter than CNN or MNIST

Also, in this example, I show how to extract the variables to print out or [save in file, if required]

### CODE REVIEW

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                        .run().get(0).expect(TFloat32.DTYPE);
+
+                System.out.println("Bias is " + biasValue.data().getFloat());
+
+                // Let's predict y for x = 10f
+                float x = 10f;
+                float predictedY = 0f;
+
+                Tensor<TFloat32> xTensor = TFloat32.scalarOf(x);
+                Tensor<TFloat32> yTensor = TFloat32.scalarOf(NO_MEANING_VALUE_TO_PUT_IN_PLACEHOLDER);
```
Comment (Craigacp) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:142]: Why do you need to feed in `yTensor`?

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            // Define gradient operators
+            List<Variable<TFloat32>> variables = new ArrayList<>();
+            variables.add(weight);
+            variables.add(bias);
+
+            Gradients gradients = tf.gradients(mse, variables);
+
+            Constant<TFloat32> alpha = tf.constant(0.2f);
+
+            ApplyGradientDescent<TFloat32> weightGradientDescent = tf.train.applyGradientDescent(weight, alpha, gradients.dy(0));
```
Comment (Craigacp) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:94]: I don't think we want to encourage people to apply gradient descent by hand, it's likely to go wrong. The other examples use the optimizers.

Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:94]: Good point, @Craigacp , will move to the optimizer usage. My point of view was to highlight the low-level things and its usage

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 1 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:142]: Need to double-check (I thought that it doesn't work withouth feeding the yTensor)

Comment (Craigacp) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:142]: If you don't request `sum` or `mse` it should work just fine without `yTensor` and if it doesn't there's a bug somewhere.

**Comment by zaleslaw** [comment]
I think, this is an another task, a lot of tutorials for tf are started
from linreg or logreg with numerical data only. Simple Mnist with reshaped
images could be a next step. When I've started with Java API it was a
missed step for me and I decide to add it. I suggest to keep both in
codebase, what do you think?



чт, 11 июн. 2020 г., 18:09 Karl Lessard <notifications@github.com>:

**Comment by Craigacp** [comment]
As it's not an mnist example why is it in the mnist package? I'm fine with having a linear regression example somewhere, but preferably under in `org.tensorflow.model.examples.regression.linear` or something similar.

**Comment by zaleslaw** [comment]
Agree, I missed the "Mnist" package name, what do you think about next
structure : the common package "training" with two subpackages "mnist" and
"linreg" for example..

Karl, need your thought too

чт, 11 июн. 2020 г., 22:47 Adam Pocock <notifications@github.com>:

**Comment by Craigacp** [comment]
I'd be fine with that, though I have a slight preference for `org.....examples.regression.linreg` and `org.....examples.classification.mnist`.

**Comment by karllessard** [comment]
I'm fine with both proposals, I'll let you pick @zaleslaw , thanks!

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import org.tensorflow.tools.Shape;
+import org.tensorflow.types.TFloat32;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+
+/**
+ * In this example the TensorFlow finds the weight and bias of the linear regression during 1 epoch,
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:35]: "the TensorFlow" -> "TensorFlow"

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ * training on observations one by one.
+ *
+ * Also, the weight and bias are extracted and printed.
+ */
+public class LinearRegressionExample {
+    /**
+     * Amount of data points.
+     */
+    private static int N = 10;
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:44]: `final` as well? Also, why are we declaring this value as a constant while other constant are mostly just hardcoded below, should we be consistent and declare constants for all of them?

I remember that a guy at Google told me before that their "rule" for this is that if a value is only used once, hardcode it, otherwise create a constant out of it.

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        for (int i = 0; i < yValues.length; i++) {
+            yValues[i] = (float) (10 * xValues[i] + 2 + 0.1 * (rnd.nextDouble() - 0.5));
+        }
+
+        try (Graph graph = new Graph()) {
+            Ops tf = Ops.create(graph);
+
+            // Define placeholders
+            Placeholder<TFloat32> X = tf.placeholder(TFloat32.DTYPE, Placeholder.shape(Shape.scalar()));
+            Placeholder<TFloat32> Y = tf.placeholder(TFloat32.DTYPE, Placeholder.shape(Shape.scalar()));
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:67]: Normal variables should be name with lowercase in Java, so `x` and `y` would be a better fit

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+            // Define variables
+            Variable<TFloat32> weight = tf.variable(Shape.scalar(), TFloat32.DTYPE);
+            Variable<TFloat32> bias = tf.variable(Shape.scalar(), TFloat32.DTYPE);
+
+            // Init variables
+            Assign<TFloat32> weightInit = tf.assign(weight, tf.constant(1f));
+            Assign<TFloat32> biasInit = tf.assign(bias, tf.constant(1f));
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:75]: You can create and assign the variables at once, by using [this endpoint](https://github.com/tensorflow/java/blob/f056482d29a97e67c167e8446bd845d402dc5647/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L7432)

So you'll endup with:
```java
Variable<TFloat32> weight = tf.variable(tf.constant(1f));
Variable<TFloat32> bias = tf.variable(tf.constant(1f));
```
The initializer can then be invoked when you run the session, check [this doc](https://github.com/tensorflow/java/blob/f056482d29a97e67c167e8446bd845d402dc5647/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Init.java#L19)

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+            ApplyGradientDescent<TFloat32> weightGradientDescent = tf.train.applyGradientDescent(weight, alpha, gradients.dy(0));
+            ApplyGradientDescent<TFloat32> biasGradientDescent = tf.train.applyGradientDescent(bias, alpha, gradients.dy(1));
+
+            try (Session session = new Session(graph)) {
+                // Initialize graph variables
+                session.runner()
+                        .addTarget(weightInit)
+                        .addTarget(biasInit)
+                        .run();
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:102]: This run can be simply replaced by `session.run(tf.init())`

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+                // Train the model on data
+                for (int i = 0; i < xValues.length; i++) {
+                    float y = yValues[i];
+                    float x = xValues[i];
+
+                    Tensor<TFloat32> xTensor = TFloat32.scalarOf(x);
+                    Tensor<TFloat32> yTensor = TFloat32.scalarOf(y);
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:110]: When allocating tensors, you need to protect their resources in a try block, like this:
```java
try (Tensor<TFloat32> xTensor = TFloat32.scalarOf(x);
    Tensor<TFloat32> yTensor = TFloat32.scalarOf(y)) {
 ...
}
```
I know it is not pretty but that is required to avoid any potential memory leaks and our examples should show this

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                            .run();
+
+                    System.out.println("Training phase");
+                    System.out.println("X is " + x + " Y is " + y);
+                }
+
+                // Extract the weight value
+                Tensor<TFloat32> weightValue = session.runner()
+                        .fetch("Variable")
+                        .run().get(0).expect(TFloat32.DTYPE);
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:126]: same thing, fetched tensors must be enclosed in a `try-with-resource` block. You probably want to fetch both values in one run as well, i.e. add `fetch("Variable_1")` in this run as well, then access it with index `[1]`.

By the way, for better readability, when you create your variable, you should give them a proper name to use for fetching, like `tf.withName("weight").variable(tf.constant(1))`. Then you fetch it with `.fetch("weight")`

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                float predictedY = 0f;
+
+                Tensor<TFloat32> xTensor = TFloat32.scalarOf(x);
+                Tensor<TFloat32> yTensor = TFloat32.scalarOf(NO_MEANING_VALUE_TO_PUT_IN_PLACEHOLDER);
+
+                predictedY = session.runner()
+                        .feed(X.asOutput(), xTensor)
+                        .feed(Y.asOutput(), yTensor)
+                        .fetch(yPredicted)
+                        .run().get(0).expect(TFloat32.DTYPE).data().getFloat();
```
Comment (karllessard) [inline tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:148]: `predictedY` should be enclosed in a `try-with-resource` block

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 4 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:44]: done

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 5 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:67]: done

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 7 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:102]: done

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 3 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:35]: done

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 2 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:94]: done

File: tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java (hunk 6 above)
Comment (zaleslaw) [reply tensorflow-examples/src/main/java/org/tensorflow/model/examples/mnist/training/LinearRegressionExample.java:75]: Great, missed the new cool initializer

**Comment by zaleslaw** [comment]
@karllessard I've fixed all your comments, great to discover a few cool things in new API, thanks!

**Comment by karllessard** [comment]
I'm merging it but I think the original proposal from @Craigacp was to move as well the current Mnist examples under `...examples.classification.mnist`. That could be done separately. Thanks @zaleslaw !

**Comment by zaleslaw** [comment]
Sorry, maybe I did something wrong with packages, will fix in the next PR

чт, 18 июн. 2020 г. в 04:01, Karl Lessard <notifications@github.com>:
